src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/DisableRuntimeMarshalling.cs

```diff
@@ -127,32 +138,237 @@ internal sealed class DisableRuntimeMarshallingAnalyzer : DiagnosticAnalyzer
         public override void Initialize(AnalysisContext context)
         {
             context.EnableConcurrentExecution();
-            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            // Report diagnostics in generated code to enable this analyzer to catch usages of
+            // delegates with non-blittable parameters that do not have [UnmanagedFunctionPointer] in source-generated interop.
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.ReportDiagnostics);
             context.RegisterCompilationStartAction(context =>
             {
                 if (context.Compilation.TryGetOrCreateTypeByMetadataName(
                     WellKnownTypeNames.SystemRuntimeCompilerServicesDisableRuntimeMarshallingAttribute,
-                    out INamedTypeSymbol? disableRuntimeMarshallingAttribute)
-                    && context.Compilation.Assembly.HasAttribute(disableRuntimeMarshallingAttribute))
+                    out INamedTypeSymbol? disableRuntimeMarshallingAttribute))
                 {
-                    var perCompilationAnalyzer = new PerCompilationAnalyzer(context.Compilation);
-                    perCompilationAnalyzer.RegisterActions(context);
+                    AutoLayoutTypeCache autoLayoutCache = new(context.Compilation);
+                    var hasDisableRuntimeMarshallingAttribute = context.Compilation.Assembly.HasAttribute(disableRuntimeMarshallingAttribute);
+                    if (hasDisableRuntimeMarshallingAttribute)
+                    {
+                        var disabledRuntimeMarshallingAssemblyAnalyzer = new DisabledRuntimeMarshallingAssemblyAnalyzer(context.Compilation, autoLayoutCache);
+                        disabledRuntimeMarshallingAssemblyAnalyzer.RegisterActions(context);
+                    }
+                    var delegateInteropUsageAnalyzer = new DelegateInteropUsageAnalyzer(context.Compilation, autoLayoutCache);
+                    delegateInteropUsageAnalyzer.RegisterActions(context, hasDisableRuntimeMarshallingAttribute);
                 }
             });
         }
 
-        private class PerCompilationAnalyzer
+        private class AutoLayoutTypeCache
```

Comment:
Given we now have multiple nested types, I think it would be more readable to move each nested type into it's own separate source file.